rclcpp/src/rclcpp/generic/typesupport_helpers.cpp

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/generic/typesupport_helpers.hpp"
+
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "ament_index_cpp/get_resources.hpp"
+#include "ament_index_cpp/get_package_prefix.hpp"
+
+#include "rcpputils/shared_library.hpp"
+
+#include "rosidl_runtime_cpp/message_type_support_decl.hpp"
+
+namespace rclcpp_generic
+{
+
+std::string get_typesupport_library_path(
+  const std::string & package_name, const std::string & typesupport_identifier)
+{
+  const char * filename_prefix;
+  const char * filename_extension;
+  const char * dynamic_library_folder;
+#ifdef _WIN32
+  filename_prefix = "";
+  filename_extension = ".dll";
+  dynamic_library_folder = "/bin/";
+#elif __APPLE__
+  filename_prefix = "lib";
+  filename_extension = ".dylib";
+  dynamic_library_folder = "/lib/";
+#else
+  filename_prefix = "lib";
+  filename_extension = ".so";
+  dynamic_library_folder = "/lib/";
+#endif
+
+  std::string package_prefix;
+  try {
+    package_prefix = ament_index_cpp::get_package_prefix(package_name);
+  } catch (ament_index_cpp::PackageNotFoundError & e) {
+    throw std::runtime_error(e.what());
+  }
+
+  auto library_path = package_prefix + dynamic_library_folder + filename_prefix +
+    package_name + "__" + typesupport_identifier + filename_extension;
+  return library_path;
+}
+
+const std::pair<std::string, std::string> extract_type_and_package(const std::string & full_type)
+{
+  std::string package_name;
+  std::string type_name;
+
+  std::tie(package_name, std::ignore, type_name) = extract_type_identifier(full_type);
+
+  return {package_name, type_name};
+}
+
+std::tuple<std::string, std::string, std::string>
+extract_type_identifier(const std::string & full_type)
```

Comment:
Same for this function, there's nothing about it really that needs to be in C++ only, and have this kind of logic duplicated in the stack can lead to inconsistencies if only one is updated for some reason in the future.